src/services/WorkPeriodService.js

```diff
@@ -0,0 +1,500 @@
+  // If one of the dates are missing then auto-calculate it
+  if (workPeriod.startDate && !workPeriod.endDate) {
+    const date = new Date(workPeriod.startDate)
+    date.setDate(date.getDate() + 6)
+    workPeriod.endDate = date
+  } else if (!workPeriod.startDate && workPeriod.endDate) {
+    const date = new Date(workPeriod.endDate)
+    date.setDate(date.getDate() - 6)
+    workPeriod.startDate = date
+  }
+
+  const resourceBooking = await helper.ensureResourceBookingById(workPeriod.resourceBookingId) // ensure resource booking exists
+  workPeriod.projectId = resourceBooking.projectId
+
+  const user = await helper.ensureUserById(resourceBooking.userId) // ensure user exists
+  workPeriod.userHandle = user.handle
+
+  workPeriod.id = uuid.v4()
+  workPeriod.createdBy = await helper.getUserId(currentUser.userId)
+
+  let created = null
+  try {
+    created = await WorkPeriod.create(workPeriod)
+  } catch (err) {
+    if (!_.isUndefined(err.original)) {
+      throw new errors.BadRequestError(err.original.detail)
+    } else {
+      throw err
+    }
+  }
+
+  await helper.postEvent(config.TAAS_WORK_PERIOD_CREATE_TOPIC, created.toJSON())
+  return created.dataValues
+}
+
+createWorkPeriod.schema = Joi.object().keys({
+  currentUser: Joi.object().required(),
+  workPeriod: Joi.object().keys({
+    resourceBookingId: Joi.string().uuid().required(),
+    startDate: Joi.workPeriodStartDate(),
+    endDate: Joi.workPeriodEndDate(),
+    daysWorked: Joi.number().integer().min(0).allow(null),
+    memberRate: Joi.number().allow(null),
+    customerRate: Joi.number().allow(null),
+    paymentStatus: Joi.paymentStatus().required()
+  }).required()
+}).required()
+
+/**
+  * Update workPeriod
+  * @param {Object} currentUser the user who perform this operation
+  * @param {String} id the workPeriod id
+  * @param {Object} data the data to be updated
+  * @returns {Object} the updated workPeriod
+  */
+async function updateWorkPeriod (currentUser, id, data) {
+  // check permission
+  await _checkUserPermissionForWriteWorkPeriod(currentUser)
+
+  const workPeriod = await WorkPeriod.findById(id)
+  const oldValue = workPeriod.toJSON()
+
+  // if resourceBookingId is provided then update projectId and userHandle
+  if (data.resourceBookingId) {
+    const resourceBooking = await helper.ensureResourceBookingById(data.resourceBookingId) // ensure resource booking exists
+    data.projectId = resourceBooking.projectId
+
+    const user = await helper.ensureUserById(resourceBooking.userId) // ensure user exists
+    data.userHandle = user.handle
+  }
+  // If one of the dates are missing then auto-calculate it
+  if (data.startDate && !data.endDate) {
+    const date = new Date(data.startDate)
+    date.setDate(date.getDate() + 6)
+    data.endDate = date
+  } else if (!data.startDate && data.endDate) {
+    const date = new Date(data.endDate)
+    date.setDate(date.getDate() - 6)
+    data.startDate = date
+  }
```

Comment:
@eisbilir let's put this logic into a reusable method and in use in both `create` and `updated` requests?

Reply:
okay